contracts/dapis/Median.sol

```diff
@@ -21,19 +21,50 @@ contract Median is Sort, Quickselect {
             if (arrayLength % 2 == 1) {
                 return array[arrayLength / 2];
             } else {
-                return
-                    (array[arrayLength / 2 - 1] + array[arrayLength / 2]) / 2;
+                assert(arrayLength != 0);
+                unchecked {
+                    return
+                        average(
+                            array[arrayLength / 2 - 1],
+                            array[arrayLength / 2]
+                        );
+                }
             }
         } else {
             if (arrayLength % 2 == 1) {
                 return array[quickselectK(array, arrayLength / 2)];
             } else {
-                (uint256 mid1, uint256 mid2) = quickselectKPlusOne(
-                    array,
-                    arrayLength / 2 - 1
-                );
-                return (array[mid1] + array[mid2]) / 2;
+                uint256 mid1;
+                uint256 mid2;
+                unchecked {
+                    (mid1, mid2) = quickselectKPlusOne(
+                        array,
+                        arrayLength / 2 - 1
+                    );
+                }
+                return average(array[mid1], array[mid2]);
             }
         }
     }
+
+    /// @notice Averages two signed integers without overflowing
+    /// @param x Integer x
+    /// @param y Integer y
+    /// @return Average of integers x and y
+    function average(int256 x, int256 y) private pure returns (int256) {
+        unchecked {
+            int256 averageRoundedDownToNegativeInfinity = (x >> 1) +
+                (y >> 1) +
+                (x & y & 1);
+            // If the average rounded down to negative infinity is negative
+            // (i.e., its 256th sign bit is set), and one of (x, y) is even and
+            // the other one is odd (i.e., the 1st bit of their xor is set),
+            // add 1 to round the average down to zero instead
+            return
+                averageRoundedDownToNegativeInfinity +
+                (int256(
+                    (uint256(averageRoundedDownToNegativeInfinity) >> 255)
```

Comment:
Solidity doesn't give us logical shift (`>>>`) and forces us to use arithmetic shift. Here, we cast the number to unsigned, do an arithmetic shift (which is the same thing as a logical shift for unsigned numbers) and back to a signed integer. So `int256(uint256(signedNumber) >> 255)` is the same thing as `signedNumber >>> 255`, it will be 1 if `signedNumber` is negative and 0 otherwise.